File: Data_Structures/Implement_Queue_With_Two_Stacks/Queue.java
package Data_Structures.Implement_Queue_With_Two_Stacks;

import java.util.LinkedList;

public class Queue<T> {

    private Stack<T> stack = new Stack<>();
    private Stack<T> backingStack = new Stack<>();

    public static void main(String... args) {

        Queue<Integer> queue = new Queue<>();

        queue.offer(1);
        queue.offer(2);
        queue.offer(3);
        queue.offer(4);
        queue.offer(5);
        queue.offer(6);

        while (!queue.isEmpty()) {
            System.out.println(queue.poll());
        }

    }

    private void addAll(Stack<T> destination, Stack<T> source) {
        while (!source.isEmpty()) {
            destination.push(source.pop());
        }
    }

    public void offer(T value) {
        stack.push(value);
    }

    public T poll() {
        addAll(backingStack, stack);
        T value = backingStack.pop();
        addAll(stack, backingStack);
        return value;
    }

    public T peek() {
        return stack.peek();
    }

    public boolean isEmpty() {
        return stack.isEmpty();
    }

}

class Stack<T> {
    private LinkedList<T> list = new LinkedList<T>();

    public void push(T value) {
        list.add(value);
    }

    public T pop() {
        return list.removeLast();
    }

    public boolean isEmpty() {
        return list.isEmpty();
    }

    public int size() {
        return list.size();
    }

    public T peek() {
        return list.getLast();
    }
}

File: Data_Structures/Implement_Queue_With_Two_Stacks/queue.py


def add_all(destination, source):
    while not source.is_empty():
        destination.push(source.pop())


class Queue:

    def __init__(self):
        self.stack = Stack()
        self.backing_stack = Stack()

    def offer(self, value):
        self.stack.push(value)

    def poll(self):
        add_all(self.backing_stack, self.stack)
        value = self.backing_stack.pop()
        add_all(self.stack, self.backing_stack)
        return value

    def is_empty(self):
        return self.stack.is_empty()


class Stack:

    def __init__(self):
        self.list = []

    def push(self, value):
        self.list.append(value)

    def pop(self):
        return self.list.pop()

    def is_empty(self):
        return len(self.list) == 0


queue = Queue()

queue.offer(1)
queue.offer(2)
queue.offer(3)
queue.offer(4)
queue.offer(5)
queue.offer(6)

while not queue.is_empty():
    print(queue.poll())
